test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp

```diff
@@ -180,14 +180,162 @@ TEST_VM(NMT, test_realloc) {
   }
 }
 
-TEST_VM_FATAL_ERROR_MSG(NMT, memory_corruption_call_stack, ".*header canary.*") {
-  if (MemTracker::tracking_level() != NMT_detail) {
-    guarantee(false, "fake message ignore this - header canary");
+#else // ASAN is enabled
+
+#define DEFINE_ASAN_TEST(test_function)                                                   \
+  TEST_VM_FATAL_ERROR_MSG(NMT_ASAN, test_function, ".*AddressSanitizer.*") {              \
+    if (MemTracker::tracking_level() > NMT_off) {                                         \
+      test_function ();                                                                   \
+    } else {                                                                              \
+      /* poisoning header/footer of memory requires NMT to be on. If off, fake assert. */ \
+      guarantee(false,                                                                    \
+                "fake message ignore this - AddressSanitizer");                           \
+    }                                                                                     \
   }
-  const size_t SIZE = 1024;
+
+static void test_write_canary() {
+  const size_t SIZE = 10;
   char* p = (char*)os::malloc(SIZE, mtTest);
-  *(p - 1) = 0;
-  os::free(p);
+  uint16_t* canary_ptr = (uint16_t*)((char*)p - sizeof(uint16_t));
+  *canary_ptr = 1;
 }
 
```

Comment:
If we simplify to poisoning the whole header (or just the canary itself) this whole section can be shrunk down to two test functions.

Otherwise, there is a large potential for simplification here. A lot of that code is duplicated.

Also note that there is a time cost involved with very fine-grained tests like these. Death tests take a long time, since they involve forking new processes, starting the JVM, letting it crash, and writing the hs_err.log file. Seeing that we basically just test different offsets in the header, I think I would reduce this to just one test, or maybe two if you want to test start and end of header (though arguably the only test really needed is the test for the canary, since it borders the user section and is the first to be hit on buffer overflow).